Add semgrep_bridge_python and semgrep_bridge_core shared libraries.

smcpeak opened the pull request:
This PR adds shared libraries allowing the semgrep Python CLI to directly invoke the OCaml core.

See `semgrep-core/src/cli/cli-bridge/bridge_design.txt` for details.

The PR also includes a few miscellaneous small changes that are unrelated other than having been done as part of the same work.

Test plan: Copying `bridge_design.txt`, revised (2022-08-24) section Testing here:

> As this work is preliminary to building a Snowflake UDF, there is not meant to be any user-visible consequence of using the bridge module.  However, its existence can be confirmed in the log file, as the command after "Running semgrep-core with command" will mention semgrep_bridge_python.so instead of semgrep-core.
>
> The main method of testing is to simply run the main test suite ("make test", etc.) with the bridge active (SEMGREP_USE_BRIDGE=require).

smcpeak committed:
core_runner.py: When failing to parse output JSON, print the exception.

This can help debug the "internal error".

smcpeak committed:
cli.py: cli: Adjust "type: ignore" for 'commands'.

At run time I can see that 'ctx.command.commands' is a dictionary.
However, that attribute is not present in the Click documentation, so
mypy is unhappy without this intermediate assignment and warning
suppression.

I do not know how to reconcile the conflict between this usage and the
documentation, nor why I see a mypy complaint that evidently the CI
build does not.

smcpeak committed:
semgrep_core.py: compute_executable_path: Catch not found exception.

In my testing, if I remove a file that path() is attempting to load,
it raises FileNotFoundError, rather than returning a path for which
is_file() fails.  Since the intent of this code is to recover, we now
catch and log the exception.

smcpeak committed:
spacegrep/scripts/show-perf: Read from /dev/urandom, not /dev/random.

/dev/random blocks when the kernel thinks it is low on entropy,
which can cause the test to hang, particularly on VMs.  This test
does not require cryptographically secure randomness, so urandom is
fine.

smcpeak committed:
qa/public_repos.py: Disable dropbox/changes.

Nested four levels deep in node_modules, the file outro.js consists
of one line, "}.call(this));", which does not parse.

smcpeak committed:
semgrep_main.py: Log the command line.

Particularly when debugging tests that invoke semgrep, it is helpful
to have the command line captured.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/returntocorp/semgrep?pullRequest=5960) <br/>All committers have signed the CLA.

smcpeak committed:
Add semgrep_bridge_python and semgrep_bridge_core shared libraries.

See semgrep-core/src/cli/cli-bridge/bridge_design.txt for details.

smcpeak commented:
I just updated the PR with changes to satisfy the code formatters and, in the github workflow actions, install `python3-config` where evidently needed.

aryx commented:
Can you add a 'Test plan:' section in your PR to explain how to test your change.
Sometimes this can be as simple as "unit tests included; make test", but in your
case it could be how how can test the new code path where the CLI communicate directly with semgrep-core.

aryx commented on semgrep-core/src/cli-bridge/bridge_design.txt:
```diff
@@ -0,0 +1,369 @@
+-------------------------------
+
+Prior to invoking the bridge module, core_runner.py forks a child
+process in which to invoke the bridge.  (Note: This is fork without
```
and fork is allowed in those map-reduce environments?

smcpeak replied:
See last paragraph in that section: "It is unknown whether ... are allowed to invoke fork."  :)  Basically, this fork is needed for things to work in normal usage, and I've got a contingency plan in case it is not allowed in a UDF.

aryx commented on semgrep-core/src/cli-bridge/bridge_design.txt:
```diff
@@ -0,0 +1,369 @@
+those enqueued.
+
+There are several ways to solve this, but doing so is outside the scope
+of the bridge module project, so I merely note this problem.
```
thx for noting this problem!

aryx reviewed (commented):
This looks great!
I didn't have the time yet to deeply review, but this looks great.

smcpeak commented:
I'll add a test plan section and look into the latest CI failures.

smcpeak committed:
tests.yml:test-core, Dockerfile: Install python3-dev.

Building semgrep_bridge_python.so needs python3-config.

smcpeak committed:
Add test_bridge_active.py.

bridge_design.txt: Add sections on envvars and testing.

smcpeak committed:
Merge remote-tracking branch 'smcpeak-repo/develop' into smcpeak-bridge2

Cli_lib.ml: Apply the changes made to Main.ml in 0e70333cec.

# Conflicts:
#	semgrep-core/src/cli/Main.ml

smcpeak committed:
install-alpine-semgrep-core: Include the bridge .so libraries.

aryx commented on cli/src/semgrep/semgrep_core.py:
```diff
@@ -35,18 +43,97 @@ def compute_executable_path(exec_name: str) -> Optional[str]:
         if cls._SEMGREP_PATH_ is None:
-            cls._SEMGREP_PATH_ = compute_executable_path("semgrep-core")
+            # Try the module first (unless disabled).
+            if not os.getenv("SEMGREP_DISABLE_BRIDGE"):
```
I'd rather make the fork/exec be the default code path, and switch to the bridge only when some environment variables are set (SEMGREP_USE_BRIDGE=1).

smcpeak replied:
Done in 4130f06e5a7807001855d20dcd2767ec028e96e8.

smcpeak committed:
semgrep_core.py: Disable the bridge unless envvar SEMGREP_USE_BRIDGE.

bridge_design.txt: Revise Enablement and Testing.  Add Portability.

Remove test_bridge_active.py: I was going to just disable it, but it
would not work as currently written since it has to launch a separate
process to go through the module loading code again.

smcpeak committed:
Merge remote-tracking branch 'smcpeak-repo/develop' into smcpeak-bridge2

brendongo commented on .github/workflows/release.yml:
```diff
@@ -143,6 +143,8 @@ jobs:
           sudo mkdir -p "$github_cache_dir"
           sudo chmod -R 777 "$github_cache_dir"
+      - name: Install python3-config
+        run: sudo apk add python3-dev
```
why was this needed?

smcpeak replied:
`python3-config` is called during the build, by `semgrep-core/src/cli-bridge/dune`.  Within this job, the `scripts/install-alpine-semgrep-core` script runs the build.  I'd be happy to put this info into a comment, but I refrained from doing so here (other than what the name says) since there aren't explanations of other steps.

brendongo commented on cli/src/semgrep/fork_subprocess.py:
```diff
@@ -0,0 +1,225 @@
+    # Although 'connect_read_pipe' is 'async', it does not really block;
+    # it will succeed or fail immediately, although notification takes a
+    # detour through the 'call_soon' queue.
+    await loop.connect_read_pipe(
```
is there a maximum capacity for the read buffer we should worry about?

smcpeak replied:
Not for `connect_read_pipe` itself, but yes once the pipe starts being used.  This is handled in `core_runner.py` by passing `limit=INPUT_BUFFER_LIMIT` in a manner parallel to the pathway that uses `asyncio.create_subprocess_exec`.  There is an existing test (`test/e2e/test_performance.py`) that fails if the limit is not set properly.

brendongo commented on cli/src/semgrep/core_runner.py:
```diff
@@ -235,35 +261,121 @@ async def _core_stderr_processor(
 
         Blocks til completion and returns exit code
         """
+        open_and_ignore("/tmp/core-runner-semgrep-BEGIN")
```
what are these for?

smcpeak replied:
My comment on `open_and_ignore` was meant to explain this:

> Attempt to open 'fname' simply so a record of having done so will be seen by 'strace'.

I was using `strace` to test which files were opened by `semgrep-core`.  To do that, I wanted a marker in the `strace` output of where `semgrep-core` starts and ends in order to easily distinguish its calls from those of the Python CLI.
